acme/writer.go

```diff
@@ -0,0 +1,283 @@
+		if err := certCfg.ManageAsync(ctx, []string{c.Domain}); err != nil {
+			cancel()
+			return err
+		}
+		c.closeCertMgr = cancel
+
+		ln = tls.NewListener(ln, tlsConfig)
+	}
+
+	authKey, found := os.LookupEnv(client.ForgeAuthEnv)
+	if found {
+		c.forgeAuthKey = authKey
+	} else {
+		fmt.Printf("NOTE: environment variable %s not set, registration is open to all peers\n", client.ForgeAuthEnv)
+	}
+
+	c.ln = ln
+
+	mux := http.NewServeMux()
+	c.nlSetup = true
+
+	// server side secret key and peerID not particularly relevant, so we can generate new ones as needed
+	sk, _, err := crypto.GenerateEd25519Key(rand.Reader)
+	if err != nil {
+		return err
+	}
+
+	authPeer := &httppeeridauth.ServerPeerIDAuth{
+		PrivKey:  sk,
+		TokenTTL: time.Hour,
+		Next: func(peerID peer.ID, w http.ResponseWriter, r *http.Request) {
+			if r.Method != http.MethodPost {
+				w.WriteHeader(http.StatusBadRequest)
+				fmt.Fprintln(w, "400 Bad Request: Only POST method is allowed.")
+				return
+			}
+			if c.forgeAuthKey != "" {
+				auth := r.Header.Get(client.ForgeAuthHeader)
+				if c.forgeAuthKey != auth {
+					w.WriteHeader(http.StatusForbidden)
+					fmt.Fprintf(w, "403 Forbidden: Missing %s header.", client.ForgeAuthHeader)
+					return
+				}
+			}
+
+			body, err := io.ReadAll(r.Body)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				_, _ = w.Write([]byte(fmt.Sprintf("error reading body: %s", err)))
+				return
+			}
+
+			typedBody := &requestBody{}
+			decoder := json.NewDecoder(bytes.NewReader(body))
+			decoder.DisallowUnknownFields()
+			if err := decoder.Decode(typedBody); err != nil {
+				w.WriteHeader(http.StatusBadRequest)
+				_, _ = w.Write([]byte(fmt.Sprintf("error decoding body: %s", err)))
+				return
+			}
+
+			// Value must be a base64url encoding of a SHA256 digest per https://datatracker.ietf.org/doc/html/rfc8555/#section-8.4
+			// It MUST NOT contain any characters outside the base64url alphabet, including padding characters ("=").
+			decodedValue, err := base64.RawURLEncoding.DecodeString(typedBody.Value)
+			if err != nil {
+				w.WriteHeader(http.StatusBadRequest)
+				_, _ = w.Write([]byte(fmt.Sprintf("error decoding value as base64url: %s", err)))
+				return
+			}
+
+			if len(decodedValue) != 32 {
+				w.WriteHeader(http.StatusBadRequest)
+				_, _ = w.Write([]byte("value is not a base64url of a SHA256 digest"))
+				return
+			}
+
+			if err := testAddresses(r.Context(), peerID, typedBody.Addresses); err != nil {
+				w.WriteHeader(http.StatusBadRequest)
+				_, _ = w.Write([]byte(fmt.Sprintf("error testing addresses: %s", err)))
+				return
+			}
+
+			const ttl = time.Hour
+			err = c.Datastore.PutWithTTL(r.Context(), datastore.NewKey(peerID.String()), []byte(typedBody.Value), ttl)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				_, _ = w.Write([]byte(fmt.Sprintf("error storing value: %s", err)))
+				return
+			}
+			w.WriteHeader(http.StatusOK)
```

Comment:
This should be `w.WriteHeader(http.StatusNoContent)` to cause a `204` since we send no content.